server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java

```diff
@@ -116,4 +205,158 @@ public void collectSearchShards(String routing, IntConsumer consumer) {
             }
         }
     }
+
+    private static class ExtractFromSource extends IndexRouting {
+        private final String indexName;
+        private final FilterPath[] include;
+
+        ExtractFromSource(int routingNumShards, int routingFactor, String indexName, List<String> routingPaths) {
+            super(routingNumShards, routingFactor);
+            this.indexName = indexName;
+            this.include = FilterPath.compile(Set.copyOf(routingPaths));
+        }
+
+        @Override
+        public int indexShard(
+            boolean idIsAutoGenerated,
+            String id,
+            @Nullable String routing,
+            XContentType sourceType,
+            BytesReference source
+        ) {
+            if (idIsAutoGenerated == false) {
+                throw new IllegalArgumentException(error("indexing with a specified id"));
+            }
+            if (routing != null) {
+                throw new IllegalArgumentException(error("indexing with a specified routing"));
+            }
+            try {
+                try (
+                    XContentParser parser = sourceType.xContent()
+                        .createParser(
+                            NamedXContentRegistry.EMPTY,
+                            DeprecationHandler.THROW_UNSUPPORTED_OPERATION,
+                            source.streamInput(),
+                            include,
```

Comment:
Do you have plans to optimize this for the 10K+ filter case? I did not dig into whether we can be smarter on the xcontent parsing/filter-path here, but current implementation looks somewhat expensive for that case.

Reply:
Yes indeed. I have plans to do more benchmarking on the case where there are a couple thousand of these filters. The `FilterPathBasedFilter` looks linear the number of filters at any particular level. I think it'd be simple enough to use a HashMap there, but I haven't looked hard yet.

That 30k number I've been throwing around is a gross estimate of the length of the text that makes up the filter.